nsxt/resource_nsxt_upgrade_prepare.go

```diff
@@ -0,0 +1,227 @@
+	client := upgrade.NewBundlesClient(connector)
+	installParam := false
+	bundleType := nsxModel.UpgradeBundleFetchRequest_BUNDLE_TYPE_PRE_UPGRADE
+	c := m.(nsxtClients)
+	userName := c.NsxtClientConfig.UserName
+	password := c.NsxtClientConfig.Password
+	upgradeBundleFetchRequest := nsxModel.UpgradeBundleFetchRequest{
+		BundleType: &bundleType,
+		Password:   &password,
+		Username:   &userName,
+		Url:        &bundleURL,
+		Version:    &version,
+	}
+	bundleID, err := client.Create(upgradeBundleFetchRequest, &installParam)
+	if err != nil {
+		return handleCreateError("NsxtUpgradePrepare", "", err)
+	}
+	err = waitForBundleUpload(m, *bundleID.BundleId)
+	if err != nil {
+		return handleCreateError("NsxtUpgradePrepare", "", err)
+	}
+	return nil
+}
+
+func acceptUserAgreement(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	acceptUserAgreement := d.Get("accept_user_agreement").(bool)
+	if !acceptUserAgreement {
+		return fmt.Errorf("To proceed with upgrade, you must accept user agreement")
+	}
+	client := eula.NewAcceptClient(connector)
+	err := client.Create()
+	if err != nil {
+		return handleCreateError("NsxtUpgradePrepare", "", err)
+	}
+	return nil
+}
+
+func executePreupgradeChecks(m interface{}) error {
+	connector := getPolicyConnector(m)
+	client := nsx.NewUpgradeClient(connector)
+	return client.Executepreupgradechecks(nil, nil, nil, nil, nil, nil)
+}
+
+func getPrecheckErrors(m interface{}, typeParam string) ([]nsxModel.UpgradeCheckFailure, error) {
+	connector := getPolicyConnector(m)
+	client := pre_upgrade_checks.NewFailuresClient(connector)
+	resultList, err := client.List(nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, &typeParam, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	return resultList.Results, nil
+}
+
+func setFailedPrechecksInSchema(d *schema.ResourceData, precheckWarnings []nsxModel.UpgradeCheckFailure) error {
+	var failedPrechecksList []map[string]interface{}
+	for _, warning := range precheckWarnings {
+		id := *warning.Id
+		message := *(warning.Message.Message)
+		elem := make(map[string]interface{})
+		elem["id"] = id
+		elem["message"] = message
+		failedPrechecksList = append(failedPrechecksList, elem)
+	}
+	return d.Set("failed_prechecks", failedPrechecksList)
+}
+
+func resourceNsxtUpgradePrepareCreate(d *schema.ResourceData, m interface{}) error {
+	id := d.Id()
+	if id == "" {
+		id = newUUID()
+	}
+	d.SetId(id)
+	// 1. Upload upgrade bundle and wait for upload to complete,
+	// UC will be automatically upgraded after this step
+	err := uploadUpgradeBundle(d, m)
+	if err != nil {
+		return handleCreateError("NsxtUpgradePrepare", id, err)
+	}
+	// 2. Accept user agreement
+	err = acceptUserAgreement(d, m)
+	if err != nil {
+		return handleCreateError("NsxtUpgradePrepare", id, err)
+	}
+	// 3. Execute pre-upgrade checks
+	err = executePreupgradeChecks(m)
+	if err != nil {
+		return handleCreateError("NsxtUpgradePrepare", id, err)
+	}
```

Comment:
I believe this code also needs to upgrade the upgrade coordinator

Reply:
I think In step 1 after we call API to upload the upgrade bundle, a API call to upgrade UC is automatically invoked.